### PR TITLE
[#182] Added line, any-line and line-count assertions for captured command output.

### LINKED ISSUE #182: Add line-oriented output assertions
## Summary

Output can only be asserted as a whole or as a substring somewhere within it. There is no way to say "the second line is exactly this", "some line matches this", or "the output has three lines". For command-line tools, whose output is inherently line-structured, that forces a brittle whole-output comparison or an over-loose substring check.

## Details

Add assertions over the `${lines[@]}` array that `bats-core` already populates:

- Assert a specific line by index, including negative indices counting from the end.
- Assert that some line matches, without pinning which one.
- Assert the line count.
- Assert an occurrence count for a pattern.

Each needs a negative counterpart. Compose these with the match modes rather than duplicating them, so a line assertion can be literal, case-sensitive or regex on the same terms as the rest of the library.

Two details worth getting right at the start:

- An out-of-range index must produce an explicit error rather than silently comparing against an empty string, which reads as an ordinary mismatch and sends the reader looking in the wrong place.
- In the failure report, print the surrounding lines with the offending one marked rather than dumping the whole stream. Marking by overwriting the indent rather than inserting a character keeps the lines aligned.

Line indices shift unless the runner is asked to preserve empty lines, so the README needs to say which option to pass and what happens without it.

## Acceptance criteria

- [ ] Index-based, any-line, line-count and occurrence-count assertions, each with a negative counterpart.
- [ ] Negative indices count from the end of the output.
- [ ] An out-of-range index fails with a message naming the index and the actual line count.
- [ ] The failure report shows the offending line in context, marked and aligned.
- [ ] Positive and negative test coverage for each assertion.
- [ ] The README documents the empty-line-preservation option and its effect on indices.

### PR DESCRIPTION
**Pull request by AlexSkrypnyk**
Closes #182

## Summary

Output could previously only be asserted as a whole (`assert_output`) or as a substring somewhere within it (`assert_output_contains`). Nothing addressed an individual line. This adds a family of assertions over the `${lines[@]}` array that bats-core's `run` already populates, covering a line at a specific index, any line regardless of position, exact line counts, and match-based occurrence counts, with every assertion composing with the library's existing match modes (literal, regex, format) and case sensitivity rather than reimplementing them.

### New `src/assert.line.bash`

- 42 public assertions across four subjects: line at index (14) - `assert_line`, `assert_line_not`, and `assert_line_[not_]{contains,matches,matches_format}[_case]`, where a negative index counts back from the end so `-1` is the last line; any line (14) - `assert_any_line` / `assert_no_line` and `assert_{any,no}_line_{contains,matches,matches_format}[_case]`; line count (2) - `assert_line_count`, `assert_line_count_not`; occurrence count (12) - `assert_line_count_[not_]{contains,matches,matches_format}[_case]`, where `not_` negates the match rather than the count, matching its meaning everywhere else in the library.
- Nine internal engines carry the logic so every public wrapper is a single line passing its anchor, polarity, mode and case sensitivity as literals: `line_resolve_index`, `line_label`, `line_context`, `line_plural`, `line_participle` and `line_validate_count` are helpers, and `line_assert_index_match`, `line_assert_index_equal`, `line_assert_any_match`, `line_assert_any_equal`, `line_assert_count` and `line_assert_count_match` are the assert engines, backed by `line_match_indices` and `line_equal_indices` for locating matching offsets.
- An out-of-range index fails with a message naming both the index and the actual line count (`Line index 5 is out of range for output with 2 lines.`) instead of comparing against an empty string, which would otherwise read as an ordinary mismatch.
- A failure prints a window of two lines either side of the offending line, with the offending line marked by overwriting the indent (`> `) rather than inserting a character, so every line in the window stays in the same column.

### `src/assert.string.bash`

- Extracted `string_match_footer` out of `string_assert_match` so the line engines can build the same "match mode / case / case-decided note" footer without duplicating that logic. `string_assert_match` keeps its exact six-argument contract and identical output.

### `load.bash`

- Sources the new `src/assert.line.bash` module.

### `tests/assert.line.bats`

- 54 tests: one per public assertion covering both the positive and the negative path, plus engine tests for index resolution (negative, zero-padded, out of range at both ends, non-integer), the context window (clamped at each end, and index padding across a width change), and a test pinning that empty lines collapse without `--keep-empty-lines` and are preserved with it.

### `README.md`

- New "Line assertions" subsection with three tables (index-based, any-line, count-based assertions), the out-of-range and context report formats, and a `--keep-empty-lines` note explaining that `run` drops empty lines unless told to keep them, which shifts every later index.
- Table of contents and the features list updated to mention line assertions.

## Before / After

```
BEFORE - no way to target, count or contextualise a single line
┌──────────────────────────────────────────────────────────────────────┐
│ run tool.sh                                                          │
│ assert_output_contains "Deleted"   # matched somewhere in the        │
│                                     # stream - which line? how many? │
│                                                                       │
│ assert_output_contains "error"     # fails by dumping the entire     │
│                                     # captured stream, however long  │
└──────────────────────────────────────────────────────────────────────┘

AFTER - a full line-assertion family, contextualised failures
┌──────────────────────────────────────────────────────────────────────┐
│ run tool.sh                                                          │
│ assert_line_contains 2 "Deleted"          # line at index            │
│ assert_any_line_matches 'Deleted [0-9]+'  # any line                 │
│ assert_line_count 4                       # how many lines total     │
│ assert_line_count_contains 2 "error"      # how many lines match     │
│                                                                       │
│ # a failing assertion now prints a small window instead of the       │
│ # whole stream, with the offending line marked in place:             │
│                                                                       │
│     0: Usage: tool.sh                                                │
│     1: Reading config                                                │
│   > 2: all good                                                      │
│     3: Done.                                                         │
└──────────────────────────────────────────────────────────────────────┘
```

## Summary by CodeRabbit

* **New Features**
  * Added line-level assertions for exact matches, text containment, regular expressions, and format patterns.
  * Supports indexed lines, negative indices, any/no matching lines, line counts, and case-sensitive or insensitive checks.
  * Added handling for empty lines, invalid inputs, and detailed contextual failure messages.

* **Documentation**
  * Expanded the README with line assertion APIs, usage details, supported options, and error behavior.

* **Bug Fixes**
  * Improved string assertion failure messages with clearer match-mode and case-sensitivity details.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -112,6 +112,105 @@ Use these after running a command with `run`.
...
+| `assert_line_count_not`              | Asserts that the number of lines differs                   |
+| `assert_line_count_contains`         | Asserts how many lines contain a string                    |
+| `assert_line_count_not_contains`     | Asserts how many lines do not contain a string             |
+| `assert_line_count_matches`          | Asserts how many lines match a regular expression          |
+| `assert_line_count_not_matches`      | Asserts how many lines do not match a regular expression   |
+| `assert_line_count_matches_format`   | Asserts how many lines match a format string               |
+| `assert_line_count_not_matches_format` | Asserts how many lines do not match a format string      |
+
+Every `contains`, `matches` and `matches_format` assertion above has a `_case` twin that matches case-sensitively - `assert_line_contains_case`, `assert_no_line_matches_case`, `assert_line_count_not_contains_case` and so on. `assert_line`, `assert_any_line`, `assert_no_line` and the two `assert_line_count` assertions compare exactly and have no twin. See [Match modes](#match-modes).
```
[the remark this answers is not in the document]
Comment (AlexSkrypnyk) [reply README.md]: Fixed in 48a2cca. Added `assert_line_not` to the list of assertions that compare exactly and have no `_case` twin.

File: src/assert.line.bash [hunk 2]
```diff
@@ -0,0 +1,1196 @@
...
+
+  # An out-of-range index would otherwise compare against an empty string,
+  # which reads as an ordinary mismatch and hides where the fault is.
+  if [ "${resolved}" -lt 0 ] || [ "${resolved}" -ge "${total}" ]; then
+    flunk "Line index ${index} is out of range for output with $(line_plural "${total}")."
+    return 1
+  fi
+
+  printf '%s\n' "${resolved}"
+}
```
[the remark this answers is not in the document]
Comment (AlexSkrypnyk) [reply src/assert.line.bash:678]: Fixed in 48a2cca. Added out-of-range cases through the public assertions: 'assert_line 4' and 'assert_line -5' in the assert_line test (past each end), and 'assert_line_contains 9' in the assert_line_contains test, each asserting the message names the index and the line count.

File: src/assert.line.bash [hunk 3]
```diff
@@ -0,0 +1,1196 @@
...
+    [ "${match_status}" -eq 0 ] && return 0
+  fi
+
+  local verb="contain"
+  local verb_third_person="contains"
+
+  if [ "${mode}" != "literal" ]; then
+    verb="match"
+    verb_third_person="matches"
+  fi
```
[the remark this answers is not in the document]
Comment (AlexSkrypnyk) [reply src/assert.line.bash:930]: Not adopting either extraction, for different reasons.

**`string_match_pattern`.** The block is only three lines of real work, and the non-format path is deliberately a direct assignment rather than a command substitution:

```bash
local pattern="${needle}"
...
pattern="$(string_format_to_regex "${needle}")"
```

Routing the literal and regex paths through a helper means the needle round-trips through `$( )`, which strips every trailing newline. `assert_output_contains $'foo\n'` currently asserts that the output contains `foo` followed by a newline; after the extraction it would silently assert only `foo`. That is a real regression in `string_assert_match`, which backs every output, stderr, string and file assertion in the library. The format path is safe because a regular expression has no meaningful trailing newline, but a helper covering only that path would not remove the duplication being flagged.

`string_match_footer` was worth extracting because it is ten lines including the second `string_match` call that computes the case-decided note, and it returns text that is already being interpolated - so there is nothing to round-trip.

**The verb pair.** It is two values, which in Bash means printing both and splitting them at each call site - more machinery than the four literal lines it would replace. It also would not fold cleanly into `string_assert_match`: the verbs there are chosen by a three-armed `case` on the anchor, where `start` and `end` produce `start with`/`starts with` and `end with`/`ends with`. Only the `*)` arm matches the line engine, so the `case` would stay and one of its three arms would call out to a helper - less readable than what is there now.
